### PR TITLE
Leaves - Raisah

### PR DESCRIPTION
**Pull request by raisahv**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. |  I wrote a custom model method in the Work model `find-spotlight` to find the work with the highest number of votes.
Describe how you approached testing that model method.  What edge cases did you come up with? | I added fixtures so that I had a number of works in my database then then created times loops to vote on those works. I then tested to see if the highest-voted work would be returned by the `find-spotlight` method. I also tested the edge case  in the event that there were no works.
What are `session` and `flash`? What is the difference between them? | Session and flash are hashes that store data that persists beyond one request. Flash data is used a single time, and once they it is displayed once they are gone. Session data is stored indefinitely (but will no longer be stored after the browser is closed)
What was one thing that you gained more clarity on through this assignment? | How to use session data in all parts of the Rails app
What is the Heroku URL of your deployed application? | https://media-ranker-r.herokuapp.com
Do you have any recommendations on how we could improve this project for the next cohort? | It seemed like a redundant project since we just made task-list and ride-share. I would consider adding some new requirements to the project that were not requirements in the last project.

### CODE REVIEW

File: app/views/homepages/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,47 @@
+
+<section class="spotlight-container">
+  <h2>Media Spotlight:</h2> 
+  <p><% spotlight = Work.find_spotlight %>
+  <%= link_to spotlight.title, work_path(spotlight) %>, by <%= spotlight.creator %>
+  </p>
+</section>
+
+<section class="top-ten-container">
```
Comment (beccaelenzil) [inline app/views/homepages/index.html.erb:9]: Consider using a partial view for these two sections to DRY up your code.

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    end
+  end
+
+  def login_form
+    @user = User.new
+  end
+
+  def login
+    username = params[:user][:username]
+    user = User.find_by(username: username)
```
Comment (beccaelenzil) [inline app/controllers/users_controller.rb:22]: Finding a logged in user is good functionality to put in a controller filter because it may be used multiple places in the future.

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    end 
+    
+    @work.update(work_params)
+    
+    redirect_to work_path(@work)
+    return
+  end
+  
+  def destroy
+    @work = Work.find_by(id: params[:id])
```
Comment (beccaelenzil) [inline app/controllers/works_controller.rb:56]: Consider putting this functionality in a controller filter.

File: app/views/works/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        </tr>
+      </thead>
+      <tbody>
+        <% Work.sort_media(:book).each do |work|  %>
+        <tr>
+          <td><%= work.votes.length%></td>
+          <td><%= link_to work.title, work_path(work) %></td>
+          <td><%= work.creator %></td>
+          <td><%= work.publication_year %></td>
+          <td><%= link_to "upvote", work_path(work), method: :post %></td>
```
Comment (beccaelenzil) [inline app/views/works/index.html.erb:27]: Looks like you have the wrong path here, and that's what's causing your bug.

File: test/models/user_test.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
+require "test_helper"
+
+describe User do
+  before do
+    @new_user = User.new(username: "Raisah")
+  end
+
+  describe "instantiation" do
+    it "can be instantiated" do
```
Comment (beccaelenzil) [inline test/models/user_test.rb:9]: Also check that it cannot be instantiated if username is nil.

File: test/models/user_test.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      expect(@new_user.valid?).must_equal true
+    end
+  end
+
+  describe "relationships" do 
+    it "can have many votes" do
+      Vote.create(user_id: @new_user.id, work_id: works(:obama).id)
```
Comment (beccaelenzil) [inline test/models/user_test.rb:16]: You could consider using a users.yml for this test.

File: test/models/vote_test.rb [hunk 7]
```diff
@@ -0,0 +1,24 @@
+require "test_helper"
+
+describe Vote do
+  before do
+    new_work = works(:miller)
```
Comment (beccaelenzil) [inline test/models/vote_test.rb:5]: Consider using fixtures.

**Comment by beccaelenzil** [comment]
# Media Ranker
## What We're Looking For



### Manual Testing

Workflow | yes / no
--- | ---
Deployed to Heroku | yes
Optional: Look and feel is similar to the original (consider styling and layout) | I see you took the opportunity to experiment with bootstrap. looks great.
**Before logging in** |
On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | there are no movies
Can go into a work's show page | yes
Verify unable to vote on a work, and get a flash message | yes
Can edit this work successfully, and get a flash message | yes
Can go to "View all media" page and see three lists of works, sorted by vote | yes -- except no movies
Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | yes
Can create a new work successfully. Note the URL for this work's show page | yes
Can delete this work successfully |yes
Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | yes
Verify that the "View all users" page lists no users (except what is produced by db seeding) | yes
**Log in** |
Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | yes
Your username is listed in "View all users" page |yes 
Verify that number of votes determines the Media Spotlight |
Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | yes
Voting on the same work twice produces an error/flash message, and there is no extra vote | upvote when logged in has a bug and redirect to "The page you were looking for doesn't exist."
**Log out** |
Logging out showed a flash message and changed the UI | yes
Logging in as a new user creates a new user | yes
Logging in as an already existing user has a specific flash message | yes

### Targeted Code Review

Area | yes / no
--- | ---
git commits were small and atomic, with useful messages | yes
Routes file uses `resources` for works |
Routes file shows intention in limiting routes for voting, log-in functionality, and users | yes
The homepage view, all media view, and new works view use semantic HTML | yes
The homepage view, all media view, and new works view use partials when appropriate | see comment
The model for media (likely named `work.rb`) `has_many` votes | yes
The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | yes
Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | see comment
Some controller, likely the `WorksController`, has a controller filter for finding a work | see comment
The `WorksController` uses strong params | yes
The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | yes

#### Targed Test Review

Area | yes / no
--- | ---
There are valid fixtures files used for users, votes, and works | yes, for works
User model has tests with sections on validations (valid and invalid) and relationships (has votes) | see comment
Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | yes
Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | yes
Work model has tests with a section on all business logic methods in the model, including their edge cases | yes

## Overall Feedback
Good work overall. Your code is well written and well tested. You make good use of session and flash, and provide your users with "polite" error handling. There is good use of works fixtures in your tests. Consider expanding your use of fixtures to the votes and users model. There is one bug with upvote which makes it so that users can't vote on the works, and you forgot to include movies in your works.
